### DISCUSSION TITLE
How to share shell profile between steps? (or: how to use nvm/rvm in steps?)

### DISCUSSION THREAD

**Discussion by scottohara** [community/community#25061] [Actions]
<p>I have a workflow consisting of two jobs:  frontend (nodejs) &amp; backend (ruby); that I’m attempting to migrate from TravisCI.</p>
<p>The frontend code uses <a href="https://github.com/nvm-sh/nvm" rel="nofollow noopener">NVM</a> with an <code>.nvmrc</code> file to specify the version of Node.JS to use.</p>
<p>The backend code uses <a href="https://rvm.io/" rel="nofollow noopener">RVM</a> with a <code>.ruby-version</code> file to specify the version of Ruby to use.</p>
<p>TravisCI build environments come with both NVM and RVM pre-installed.</p>
<p>The version of Ruby specified in <code>.ruby-version</code> is automatically detected/installed when the Travis job runs; and a simple <code>nvm install</code> command in the preinstall stage is all that is required to have Travis detect/install the appropriate version of Node.JS.</p>
<p>In Github Actions virtual environments, default versions of both Ruby &amp; Node.JS are pre-installed, however version managers such as NVM/RVM are not.</p>
<p>The ‘official’ way to install a different Ruby/Node version in Github Actions appears to via <a href="https://github.com/actions/setup-ruby" rel="nofollow noopener">actions/setup-ruby</a> and <a href="https://github.com/actions/setup-node" rel="nofollow noopener">actions/setup-node</a>.</p>
<p>Unfortunately, <a href="https://github.com/actions/setup-node/issues/32" rel="nofollow noopener"><code>setup-node</code> doesn’t undersand <code>.nvmrc</code> files yet</a>, and <a href="https://github.com/actions/setup-ruby/issues/31" rel="nofollow noopener"><code>setup-ruby</code> doesn’t understand <code>.ruby-version</code> files yet</a> (so the specfied runtime versions would need to be duplicated in the workflow).</p>
<p>To emulate TravisCI, I’m trying to install both NVM and RVM as the first steps in their respective jobs, so that the jobs can use the .nvmrc and .ruby-version files to detect &amp; install the appropriate runtime versions.</p>
<p>The challenge I have is that both NVM and RVM expect to be sourced into the shell after installation, e.g.</p>
<pre><code>echo "source $NVM_DIR/nvm.sh" &gt;&gt; "$HOME/.bash_profile"
...

echo "source $HOME/.rvm/scripts/rvm" &gt;&gt; "$HOME/.bash_profile"
...
</code></pre>
<p>As I’ve discovered, each <code>run</code> command in a job is a separate shell process; and the default shell command on the Ubuntu environments is:</p>
<pre><code>bash --noprofile --norc -eo pipefail {0}
</code></pre>
<p>(explicitly instructed to not read from <code>.bash_profile</code> and <code>.bashrc</code> files).</p>
<p>I tried adding an explicit <code>shell: bash {0}</code> to each step, but this does not seem to have made any difference.</p>
<p>Below is an example workflow.</p>
<p>For nvm: the <code>--install</code> argument when sourcing into the shell tells nvm to auto-install &amp; switch to the version of node found in the .nvmrc file</p>
<p>For rvm: the <code>rvm_install_on_use_flag</code> environment variable tells rvm to auto-install &amp; switch to the version of ruby found in the .ruby-version file.</p>
<pre><code>frontend:
    runs-on: ubuntu-latest

    steps:
    - name: Checkout repository
      uses: actions/checkout@v1

- name: Check versions (before)
shell: bash {0}
run: |
node --version
npm --version
    - name: Install nvm
      shell: bash {0}
      run: |
        curl -o- https://raw.githubusercontent.com/nvm-sh/nvm/v0.34.0/install.sh | bash
        export NVM_DIR="$HOME/.nvm"
        echo "source $NVM_DIR/nvm.sh --install" &gt;&gt; "$HOME/.bash_profile"

    - name: Check versions (after)
      shell: bash {0}
      run: |
        node --version
        npm --version

  backend:
    runs-on: ubuntu-latest

    steps:
    - name: Checkout repository
      uses: actions/checkout@v1

- name: Install rvm
      shell: bash {0}
      run: |
        gpg --keyserver hkp://pool.sks-keyservers.net --recv-keys 409B6B1796C275462A1703113804BB82D39DC0E3 7D2BAF1CF37B13E2069D6956105BD0E739499BDB
        curl -sSL https://get.rvm.io | bash -s stable
        export rvm_install_on_use_flag=1
        echo "source $HOME/.rvm/scripts/rvm" &gt;&gt; "$HOME/.bash_profile"

    - name: Check versions (after)
      shell: bash {0}
      run: |
        rvm --version
        ruby --version
        gem --version
        bundler --version
</code></pre>
<p>For the frontend job, I had expected <code>node --version</code> to be v10.16.3 in the ‘before’ step; and v12.1.0 in the ‘after’ step;  but it is v10.16.3 in both.</p>
<p>It seems that the changes to <code>.bash_profile</code> are not seen by other steps?  (for what it’s worth, I also tried <code>.bashrc</code> instead…to no avail).</p>
<p>Is there a way that I can have the NVM/RVM scripts sourced into all subsequent shell processes in a job?</p>
<p>(also, having to repeat <code>shell: bash {0}</code> in every step is pretty annoying).</p>

**Comment by dakale** [comment, marked as answer]
<p>You were close: You can actually just add <code>shell: bash -l {0}</code> (the -l was missing from yours) for steps where you need to source the bash_profile. See: <a href="https://www.gnu.org/software/bash/manual/html_node/Bash-Startup-Files.html" rel="nofollow noopener">https://www.gnu.org/software/bash/manual/html_node/Bash-Startup-Files.html</a>, </p>
<p>I am aso assuming that install script primarily exports/updates the PATH to include the nvm install location, so youd only need to add the shell option for steps that need to exec node</p>
<p>The default was to invoke bash non-interactive and non-login since semantically thats closest to what we are doing- just running scripts, and I wanted to avoid adding potential side effects, unexpected behavior, re-running the same statup files repeatedly for each step, etc. The shell option was intended to allow you to manage these types of scenarios as necessary. Its not really the expected default behavior since we have setup actions that download/install tools and modify the PATH, etc, so that the tools installed are available everywhere not just from bash steps that pass down the environment, PATH is tracked by the runner across steps, etc</p>

**Comment by normanlolx** [comment]
<p>To have a working sample snippet. Adding a local Composer’s bin to the path. The Drupal step would fail without <span>shell: </span>bash -l {0}.</p>
<pre><code>jobs:  
 build:  
 name: Build  
runs-on: ubuntu-latest  
  
steps:  
 - uses: actions/checkout@v2  
  
- name: Composer  
run: |  
 composer --version  
 composer validate  
 composer install -n  
 echo "export PATH=$PATH:$GITHUB\_WORKSPACE/vendor/bin" \&gt;\&gt; ~/.bash\_profile  
 source ~/.bash\_profile  
 drush --version  
  
- name: Drupal  
shell: bash -l {0}  
run: |  
 drush --version
</code></pre>

**Comment by dreinhardt89** [comment]
Putting this here as it's an issue I ran into. If you don't also set -e the job status may not update correctly (this may only happen if the script is run in a composite action though).

i.e. ```shell: bash --norc -l -e -o pipefail {0}```
